File: schema-validation/src/delomaten_schema_validation/validate_config.py
from asdf.schema import ValidationError, validate # type: ignore
from dataclasses import dataclass
from enum import Enum
import os
from rich.color import Color
from rich.console import Console
from rich.highlighter import Highlighter
from rich.style import Style
import sys
import textwrap
from typing import Any # type: ignore
import yaml


# Highlights all text in red
class ErrorHighlighter(Highlighter):
    def highlight(self, text):
        text.stylize(Style(color=Color.from_rgb(255,0,0)), 0, len(text))

class SuccessHighlighter(Highlighter):
    def highlight(self, text):
        text.stylize(Style(color=Color.from_rgb(0,255,0)), 0, len(text))

# Force terminal so that colours are printed in the github action runner
errorConsole = Console(highlighter=ErrorHighlighter(), force_terminal=True)
console = Console(highlighter=SuccessHighlighter(), force_terminal=True)

# data model for delomaten configuration file
class PseudoOperation(Enum):
    PSEUDO = "PSEUDO"
    DEPSEUDO = "DEPSEUDO"
    REDACT = "REDACT"

class EncryptionAlgorithm(Enum):
    PAPIS_COMPATIBLE = "papis_compatible"
    SID_MAPPING = "sid_mapping"

@dataclass
class PseudoTask:
    name: str
    columns: list[str]
    pseudo_operation: PseudoOperation
    encryption_algorithm: EncryptionAlgorithm
    encryption_args: list[str] | None = None

@dataclass
class DelomatenConfig:
    shared_bucket: str
    source_folder: str
    destination_folder: str
    memory_size: int
    pseudo: list[PseudoTask]
    output_columns: list[str] | None = None

def load_delomaten_config(file_path: str) -> DelomatenConfig:
    with open(file_path, "r") as stream:
        return DelomatenConfig(**yaml.safe_load(stream))

def validate_config(environment: str, folder: str, directory_path: str, shared_buckets_path: str) -> None:
    """
    Validate if the config.yaml under the @directory_path is valid.
    Only uses @environment and @folder arguments to print more informative logs.

    Parameters:
        @environment: The dapla team environment the source belongs to i.e. 'uh-varer-prod' or 'play-obr-test'
        @folder: The statistics product folder which contains the configuration file i.e. 'ledstill' or 'sykefra'
        @directory_path: The filepath to the config.yaml file containing 'delomaten' configuration
        @shared_buckets_path: The filepath to the shared-buckets iam.yaml file containing the teams' shared-buckets

    """
    if not os.path.exists(directory_path):
        errorConsole.print(f'\n\nThe given directory path "{directory_path}" does not exist')
        sys.exit(1)

    config_data_path: str = os.path.join(directory_path, 'config.yaml')

    # We expect to find a config.yaml file in the directory
    if not os.path.exists(config_data_path):
        errorConsole.print(
            f'\n\nNo "config.yaml" file exists in the product source folder "{os.path.join(environment, folder)}"\n\n'
        )
        sys.exit(1)

    if not os.path.exists(shared_buckets_path):
        errorConsole.print(f'\n\nThe given shared-buckets path "{shared_buckets_path}" does not exist')
        sys.exit(1)

    # with open(config_data_path) as stream:
    #     config_data: dict[str, str | list[str] | list[dict[str, str | list[str]]]] = yaml.safe_load(stream.read())

    config_data = load_delomaten_config(config_data_path)

    with open('./config_schema.yaml') as stream:
        config_schema: dict[str, Any] = yaml.safe_load(stream.read())

    contextual_path = os.path.join(environment, folder, 'config.yaml')
    try:
        validate(config_data, schema=config_schema)
    except ValidationError as e:
        errorConsole.print(f'\n\nThe delomaten configuration file "{contextual_path}" is invalid:\n\n{e}\n\n')
        sys.exit(1)

    with open(shared_buckets_path) as stream:
        shared_buckets_data: dict[str, dict[str, dict[str, list[str]]]] = yaml.safe_load(stream.read())

    # If the shared bucket name specified in the config.yaml doesn't exist in the dapla team, report an error
    if config_data.shared_bucket not in shared_buckets_data['buckets'].keys():
        errorConsole.print(textwrap.dedent(
            f"""


            In the configuration file "{contextual_path}" in the field "shared_bucket" the provided bucket "{config_data.shared_bucket}" does not exist.

            Existing shared buckets for {environment}:
              {"\n".join(['- ' + bucket for bucket in shared_buckets_data['buckets'].keys()])}


            """
        ))
        sys.exit(1)

    pseudo_targeted_columns: set[str] = set(
        [column for task in config_data.pseudo for column in task.columns]
    )
    if config_data.output_columns is not None and len(diff := pseudo_targeted_columns - set(config_data.output_columns)) > 0:
        errorConsole.print(textwrap.dedent(
            f"""


               In the configuration file "{contextual_path}" in the field "output_columns" not all columns
               targeted by pseudo operations are listed in the "output_columns".

               The missing columns are:
                 {"\n".join(['- ' + column for column in diff])}


               """
        ))
        sys.exit(1)


    pseudo_task_columns_uniquely_targeted(config_data.pseudo)

    console.print(f'The "{contextual_path}" configuration was successfully validated!')

def pseudo_task_columns_uniquely_targeted(pseudo_tasks: list[PseudoTask]) -> None:
    """Ensure that the pseudo task columns are only targeted once."""
    assoc_dict: dict[str, list[str]] = (
        {task.name : task.columns for task in pseudo_tasks}
    )

    for pseudo_task_name_a, targeted_columns_a in assoc_dict.items():
        for pseudo_task_name_b, targeted_columns_b in {k:v for k,v in assoc_dict.items() if k != pseudo_task_name_a }.items():
            tcs_a: set[str] = set(targeted_columns_a)
            tcs_b: set[str] = set(targeted_columns_b)
            if len(overlapping_cols := tcs_a & tcs_b) > 0:
                errorConsole.print(
                    f"""
                    The pseudo tasks "{pseudo_task_name_a}" and "{pseudo_task_name_b}" target overlapping columns.

                    Overlapping columns:
                      {"\n".join([f'- "{column}"' for column in overlapping_cols])}
                    """
                )
                sys.exit(1)


if __name__ == '__main__':
  [environment, folder, directory_path, shared_buckets_file_path, *rest] = sys.argv[1:]
  validate_config(environment, folder, directory_path, shared_buckets_file_path)

File: schema-validation-sc/src/validateConfig.scala
//> using scala 3.6.3
//> using dep org.typelevel::cats-core:2.13.0
//> using dep io.circe::circe-yaml:1.15.0
//> using dep io.circe::circe-generic:0.14.10
//> using dep io.circe::circe-parser:0.14.10
//> using dep com.networknt:json-schema-validator:1.5.5
//> using dep ch.qos.logback:logback-classic:1.5.16
//> using files types.scala utils.scala
package validate

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.networknt.schema.serialization.JsonNodeReader
import com.networknt.schema.SpecVersion.VersionFlag
import com.networknt.schema.*
import com.schemavalidation.types.{given, *}
import com.schemavalidation.utils.loadConfig
import io.circe.*
import io.circe.syntax.*
import io.circe.yaml
import java.nio.file.{Files, Path, Paths}
import scala.collection.immutable.*
import scala.io.Source
import scala.jdk.CollectionConverters.*

/** Validate if the config.yaml under the @directory_path is valid. Only uses
  * \@environment and @folder arguments to print more informative logs.
  *
  * @param environment
  *   the dapla team environment the source belongs to i.e. 'uh-varer-prod' or
  *   'play-obr-test'
  * @param folder
  *   the statistics product folder which contains the configuration file i.e.
  *   'ledstill' or 'sykefra'
  * @param directory_path
  *   the filepath to the config.yaml file containing 'delomaten' configuration
  * @param shared_buckets_path
  *   the filepath to the shared-buckets iam.yaml file containing the teams'
  *   shared-buckets
  * @return
  *   Unit
  */
@main def validateConfig(
    environment: String,
    folder: String,
    directoryPath: String,
    sharedBucketsPath: String
): Unit =

  if !Files.exists(Paths.get(directoryPath)) then
    println(
      s"The given directory path ${directoryPath} does not exist".red.newlines
    )
    System.exit(1)

  val configDataPath: Path = Paths.get(directoryPath, "config.yaml")

  if !Files.exists(configDataPath) then
    val context = Paths.get(environment, folder).toString()
    println(
      s"No 'config.yaml' file exists in the product source folder '${context}'".red.newlines
    )
    System.exit(1)

  if !Files.exists(Paths.get(sharedBucketsPath)) then
    println(
      s"The given shared-buckets path '${sharedBucketsPath}' does not exist".red.newlines
    )
    System.exit(1)

  validateConfigSchema(configDataPath) match
    case Left(errMessages) =>
      val contextualPath = Paths.get(environment, folder, "config.yaml")
      println(
        s"The delomaten configuration file '${contextualPath}' is invalid:\n\n${errMessages.mkString("\n")}".red.newlines
      )
    case Right(_) => ()

  val delomaten: DelomatenConfig = loadConfig(configDataPath)
  val sharedBuckets: SharedBuckets = loadConfig(Paths.get(sharedBucketsPath))

  // If the shared bucket specified in the config.yaml doesn't exist in the dapla team, report an error
  if !sharedBuckets.contains(delomaten.sharedBucket) then
    println(s"""
      |In the configuration file "${configDataPath}" in the field "shared_bucket" the provided bucket "${delomaten.sharedBucket}" does not exist.

      |Existing shared buckets for ${environment}:
      |  ${sharedBuckets.map("- " + _).mkString("\n")}
    """.stripMargin.red.newlines)
    System.exit(1)

  val pseudoTargetedColumns: Set[String] =
    delomaten.pseudo.flatMap(_.columns).toSet

  delomaten.outputColumns match
    case Some(columns) if (pseudoTargetedColumns &~ columns.toSet).size > 0 =>
      val diff = pseudoTargetedColumns &~ columns.toSet
      println(s"""
          |In the configuration file '${configDataPath}' in the field 'output_columns'
          |not all columns targeted by pseudo operations are listed in the 'output_columns'.

          |The missing columns are:
          |  ${diff.map("- " + _).mkString("\n")}
        """.stripMargin.red.newlines)
      System.exit(1)
    case _ => ()

  pseudoTaskColumnsUniquelyTargeted(delomaten.pseudo)

  println(
    s"The '${configDataPath}' configuration was successfully validated!".green.newlines
  )

// Ensure that the pseudo task columns are only targeted once.
def pseudoTaskColumnsUniquelyTargeted(pseudoTasks: List[PseudoTask]): Unit =
  val assocMap: Map[String, List[String]] =
    pseudoTasks.map(task => task.name -> task.columns).toMap

  for (taskName, targetedColumns) <- assocMap do
    val remainingMap = assocMap - taskName
    for (taskNameB, targetedColumnsB) <- remainingMap do
      val overlappingColumns = Set(targetedColumns) & Set(targetedColumnsB)
      if overlappingColumns.size > 0 then
        println(
          s"""
            |The pseudo tasks '${taskName}' and '${taskNameB}' target overlapping columns.

            |Overlapping columns:
            |  ${overlappingColumns.map("- " + _).mkString("\n")}
          """.stripMargin.red.newlines
        )
        System.exit(1)

def validateConfigSchema(filepath: Path): Either[Set[ValidationMessage], Unit] =
  val schemaData: String =
    Source.fromFile("./config_schema.yaml").getLines().mkString("\n")
  val inputData: String = Source
    .fromFile(filepath.toFile())
    .getLines()
    .mkString("\n")

  val yamlMapper: ObjectMapper = ObjectMapper(YAMLFactory())
  val jsonNodeReader: JsonNodeReader =
    JsonNodeReader.builder().yamlMapper(yamlMapper).build()
  val config: SchemaValidatorsConfig =
    SchemaValidatorsConfig.builder().errorMessageKeyword("errorMessage").build()
  val schema: JsonSchema = JsonSchemaFactory
    .getInstance(
      VersionFlag.V202012,
      builder => builder.jsonNodeReader(jsonNodeReader).build()
    )
    .getSchema(schemaData, InputFormat.YAML, config)
  val messages: Set[ValidationMessage] = schema
    .validate(
      inputData,
      InputFormat.YAML,
      (executionContext) =>
        executionContext.getExecutionConfig().setFormatAssertionsEnabled(true)
    )
    .asScala
    .toSet

  if messages.size == 0 then Right(())
  else Left(messages)
